src/http/proxy_server.rs

```diff
@@ -27,8 +38,40 @@ pub(crate) async fn start(addr: SocketAddr) {
         .expect("An error occured when starting the server");
 }
 
-async fn request_handler(_req: Request<Body>) -> Result<Response<Body>, Infallible> {
-    println!("{} {}", _req.method(), _req.uri());
+async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Infallible> {
+    let method = req.method();
+
+    println!("{} {}", method, req.uri());
+
+    let fpm_url = format!("{}{}", fpm_url, req.uri());
+
+    let mut surf_response = match method.as_str() {
+        "GET" => surf::get(fpm_url).await,
+        "POST" => surf::post(fpm_url).await,
+        "PUT" => surf::put(fpm_url).await,
+        "PATCH" => surf::patch(fpm_url).await,
+        "HEAD" => surf::head(fpm_url).await,
+        "OPTIONS" => surf::options(fpm_url).await,
+        "TRACE" => surf::trace(fpm_url).await,
+        "CONNECT" => surf::connect(fpm_url).await,
+        "DELETE" => surf::delete(fpm_url).await,
+        _ => panic!(format!("Unsupported method {}", method)),
+    }
+    .unwrap();
```

Comment:
You can use the `?` operator instead of `unwrap`. And `Infallible` as the error type is probably not what you want here (it's meant for an operation that can't fail).

Reply:
There are tons of other issues if I do that, it's like a neverending waterfall of errors as long as I change one single line of code 🤣 

Like if I do that:

```diff

-async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Infallible> {
+async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, std::error::Error> {

    let method = req.method();

// ...

        "DELETE" => surf::delete(fpm_url).await,
        _ => panic!(format!("Unsupported method {}", method)),
-    }
-    .unwrap();
+    }?;
```

I get another bunch compilation errors:

<details>
<summary>Logs:</summary>

```
C:/Users/Pierstoval/.cargo/bin/cargo.exe build --color=always --all --all-targets
   Compiling rymfony v0.1.0 (E:\dev\www\rymfony)
warning: trait objects without an explicit `dyn` are deprecated
  --> src\http\proxy_server.rs:42:88
   |
42 | async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                                        ^^^^^ help: use `dyn`: `dyn Error`
   |
   = note: `#[warn(bare_trait_objects)]` on by default

warning: trait objects without an explicit `dyn` are deprecated
  --> src\http\proxy_server.rs:42:88
   |
42 | async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                                        ^^^^^ help: use `dyn`: `dyn Error`
   |
   = note: `#[warn(bare_trait_objects)]` on by default

error[E0277]: the size for values of type `dyn std::error::Error` cannot be known at compilation time
  --> src\http\proxy_server.rs:49:29
   |
49 |       let mut surf_response = match method.as_str() {
   |  _____________________________^
50 | |         "GET" => surf::get(fpm_url).await,
51 | |         "POST" => surf::post(fpm_url).await,
52 | |         "PUT" => surf::put(fpm_url).await,
...  |
59 | |         _ => panic!(format!("Unsupported method {}", method)),
60 | |     }?;
   | |______^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `dyn std::error::Error`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required by `std::ops::Try::from_error`

error[E0277]: `?` couldn't convert the error to `dyn std::error::Error`
  --> src\http\proxy_server.rs:60:6
   |
60 |     }?;
   |      ^ the trait `std::convert::From<std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>` is not implemented for `dyn std::error::Error`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required by `std::convert::From::from`

error[E0308]: mismatched types
  --> src\http\proxy_server.rs:71:5
   |
71 | /     Result::<_, Infallible>::Ok(
72 | |         Response::builder()
73 | |             .status(StatusCode::from_bytes(status.as_slice()).unwrap())
74 | |             .body(Body::from(surf_response.body_bytes().await.unwrap()))
75 | |             .unwrap()
76 | |     )
   | |_____^ expected trait object `dyn std::error::Error`, found enum `std::convert::Infallible`
   |
   = note: expected enum `std::result::Result<_, (dyn std::error::Error + 'static)>`
              found enum `std::result::Result<_, std::convert::Infallible>`

error[E0277]: the size for values of type `dyn std::error::Error` cannot be known at compilation time
  --> src\http\proxy_server.rs:49:29
   |
49 |       let mut surf_response = match method.as_str() {
   |  _____________________________^
50 | |         "GET" => surf::get(fpm_url).await,
51 | |         "POST" => surf::post(fpm_url).await,
52 | |         "PUT" => surf::put(fpm_url).await,
...  |
59 | |         _ => panic!(format!("Unsupported method {}", method)),
60 | |     }?;
   | |______^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `dyn std::error::Error`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required by `std::ops::Try::from_error`

error[E0277]: `?` couldn't convert the error to `dyn std::error::Error`
  --> src\http\proxy_server.rs:60:6
   |
60 |     }?;
   |      ^ the trait `std::convert::From<std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>>` is not implemented for `dyn std::error::Error`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required by `std::convert::From::from`

error[E0277]: the size for values of type `(dyn std::error::Error + 'static)` cannot be known at compilation time
   --> src\http\proxy_server.rs:42:95
    |
42  |   async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
    |  _______________________________________________________________________________________________^
43  | |     let method = req.method();
44  | |
45  | |     println!("{} {}", method, req.uri());
...   |
76  | |     )
77  | | }
    | |_^ doesn't have a size known at compile-time
    | 
   ::: E:\.cache\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\src\libcore\result.rs:247:20
    |
247 |   pub enum Result<T, E> {
    |                      - required by this bound in `std::result::Result`
    |
    = help: the trait `std::marker::Sized` is not implemented for `(dyn std::error::Error + 'static)`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>

error[E0271]: type mismatch resolving `<impl std::future::Future as std::future::Future>::Output == std::result::Result<http::response::Response<_>, _>`
  --> src\http\proxy_server.rs:30:43
   |
30 |     let http_server = Server::bind(&addr).serve(make_service_fn);
   |                                           ^^^^^ expected opaque type, found enum `std::result::Result`
...
42 | async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                 ----------------------------- the `Output` of this `async fn`'s expected opaque type
   |
   = note: expected opaque type `impl std::future::Future`
                     found enum `std::result::Result<http::response::Response<_>, _>`
   = note: required because of the requirements on the impl of `tower_service::Service<http::request::Request<hyper::body::body::Body>>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<hyper::body::body::Body>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`

error[E0277]: the trait bound `hyper::common::exec::Exec: hyper::common::exec::H2Exec<impl std::future::Future, _>` is not satisfied
  --> src\http\proxy_server.rs:37:5
   |
37 | /     http_server
38 | |         .await
   | |______________^ the trait `hyper::common::exec::H2Exec<impl std::future::Future, _>` is not implemented for `hyper::common::exec::Exec`
   |
   = help: the following implementations were found:
             <hyper::common::exec::Exec as hyper::common::exec::H2Exec<F, B>>
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::conn::spawn_all::NewSvcTask<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>`
   = note: required because of the requirements on the impl of `hyper::common::exec::NewSvcExec<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>` for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::Server<hyper::server::tcp::AddrIncoming, hyper::service::make::MakeServiceFn<[closure@src\http\proxy_server.rs:21:43: 28:6 fpm_url:_]>>`

error[E0271]: type mismatch resolving `<impl std::future::Future as std::future::Future>::Output == std::result::Result<http::response::Response<_>, _>`
  --> src\http\proxy_server.rs:37:5
   |
37 | /     http_server
38 | |         .await
   | |______________^ expected opaque type, found enum `std::result::Result`
...
42 |   async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                   ----------------------------- the `Output` of this `async fn`'s expected opaque type
   |
   = note: expected opaque type `impl std::future::Future`
                     found enum `std::result::Result<http::response::Response<_>, _>`
   = note: required because of the requirements on the impl of `tower_service::Service<http::request::Request<hyper::body::body::Body>>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<hyper::body::body::Body>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::common::exec::NewSvcExec<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>` for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::Server<hyper::server::tcp::AddrIncoming, hyper::service::make::MakeServiceFn<[closure@src\http\proxy_server.rs:21:43: 28:6 fpm_url:_]>>`

error: aborting due to 7 previous errors; 1 warning emitted

Some errors have detailed explanations: E0271, E0277, E0308.
For more information about an error, try `rustc --explain E0271`.
error[E0308]: mismatched types
  --> src\http\proxy_server.rs:71:5
   |
71 | /     Result::<_, Infallible>::Ok(
72 | |         Response::builder()
73 | |             .status(StatusCode::from_bytes(status.as_slice()).unwrap())
74 | |             .body(Body::from(surf_response.body_bytes().await.unwrap()))
75 | |             .unwrap()
76 | |     )
   | |_____^ expected trait object `dyn std::error::Error`, found enum `std::convert::Infallible`
   |
   = note: expected enum `std::result::Result<_, (dyn std::error::Error + 'static)>`
              found enum `std::result::Result<_, std::convert::Infallible>`

error[E0277]: the size for values of type `(dyn std::error::Error + 'static)` cannot be known at compilation time
   --> src\http\proxy_server.rs:42:95
    |
42  |   async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
    |  _______________________________________________________________________________________________^
43  | |     let method = req.method();
44  | |
45  | |     println!("{} {}", method, req.uri());
...   |
76  | |     )
77  | | }
    | |_^ doesn't have a size known at compile-time
    | 
   ::: E:\.cache\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\src\libcore\result.rs:247:20
    |
247 |   pub enum Result<T, E> {
    |                      - required by this bound in `std::result::Result`
    |
    = help: the trait `std::marker::Sized` is not implemented for `(dyn std::error::Error + 'static)`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>

error: could not compile `rymfony`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0271]: type mismatch resolving `<impl std::future::Future as std::future::Future>::Output == std::result::Result<http::response::Response<_>, _>`
  --> src\http\proxy_server.rs:30:43
   |
30 |     let http_server = Server::bind(&addr).serve(make_service_fn);
   |                                           ^^^^^ expected opaque type, found enum `std::result::Result`
...
42 | async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                 ----------------------------- the `Output` of this `async fn`'s expected opaque type
   |
   = note: expected opaque type `impl std::future::Future`
                     found enum `std::result::Result<http::response::Response<_>, _>`
   = note: required because of the requirements on the impl of `tower_service::Service<http::request::Request<hyper::body::body::Body>>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<hyper::body::body::Body>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`

error[E0277]: the trait bound `hyper::common::exec::Exec: hyper::common::exec::H2Exec<impl std::future::Future, _>` is not satisfied
  --> src\http\proxy_server.rs:37:5
   |
37 | /     http_server
38 | |         .await
   | |______________^ the trait `hyper::common::exec::H2Exec<impl std::future::Future, _>` is not implemented for `hyper::common::exec::Exec`
   |
   = help: the following implementations were found:
             <hyper::common::exec::Exec as hyper::common::exec::H2Exec<F, B>>
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::conn::spawn_all::NewSvcTask<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>`
   = note: required because of the requirements on the impl of `hyper::common::exec::NewSvcExec<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>` for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::Server<hyper::server::tcp::AddrIncoming, hyper::service::make::MakeServiceFn<[closure@src\http\proxy_server.rs:21:43: 28:6 fpm_url:_]>>`

error[E0271]: type mismatch resolving `<impl std::future::Future as std::future::Future>::Output == std::result::Result<http::response::Response<_>, _>`
  --> src\http\proxy_server.rs:37:5
   |
37 | /     http_server
38 | |         .await
   | |______________^ expected opaque type, found enum `std::result::Result`
...
42 |   async fn handle_request(fpm_url: String, req: Request<Body>) -> Result<Response<Body>, Error> {
   |                                                                   ----------------------------- the `Output` of this `async fn`'s expected opaque type
   |
   = note: expected opaque type `impl std::future::Future`
                     found enum `std::result::Result<http::response::Response<_>, _>`
   = note: required because of the requirements on the impl of `tower_service::Service<http::request::Request<hyper::body::body::Body>>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::service::http::HttpService<hyper::body::body::Body>` for `hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>`
   = note: required because of the requirements on the impl of `hyper::common::exec::NewSvcExec<hyper::server::tcp::addr_stream::AddrStream, impl std::future::Future, hyper::service::util::ServiceFn<[closure@src\http\proxy_server.rs:23:31: 25:10 fpm_url:_], hyper::body::body::Body>, hyper::common::exec::Exec, hyper::server::conn::spawn_all::NoopWatcher>` for `hyper::common::exec::Exec`
   = note: required because of the requirements on the impl of `std::future::Future` for `hyper::server::Server<hyper::server::tcp::AddrIncoming, hyper::service::make::MakeServiceFn<[closure@src\http\proxy_server.rs:21:43: 28:6 fpm_url:_]>>`

error: aborting due to 7 previous errors; 1 warning emitted

Some errors have detailed explanations: E0271, E0277, E0308.
For more information about an error, try `rustc --explain E0271`.
error: could not compile `rymfony`.

To learn more, run the command again with --verbose.

Process finished with exit code 101

```
</details>

Maybe I wrongly followed your advice? 😕

Reply:
You can't use `Error` in the return type as that's a trait. If you really need to return any error type, you have to use `Box<dyn Error + Send>`. Considering you're building a binary and not a library, the usual recommendation would be to use the `anyhow` crate (basically converting to a wrapper error type).